src/main/java/org/broadinstitute/hellbender/utils/reference/FastaReferenceWriter.java

```diff
@@ -0,0 +1,665 @@
+     * @throws IllegalArgumentException if {@code fastaFile} is {@code null} or {@code basesPerLine} is 0 or negative.
+     * @throws IOException if such exception is thrown when accessing the output path resources.
+     */
+    public FastaReferenceWriter(final Path fastaFile, final int basesPerLine, final boolean makeFaiOutput,
+                                final boolean makeDictOutput)
+        throws IOException
+    {
+        this(Utils.nonNull(fastaFile, "the input fasta-file cannot be null"),
+                basesPerLine,
+                defaultFaiFile(makeFaiOutput, fastaFile),
+                defaultDictFile(makeDictOutput, fastaFile));
+    }
+
+    /**
+     * Creates a reference FASTA file writer.
+     * <p>
+     *     The default bases-per-line is set to {@link #DEFAULT_BASES_PER_LINE}.
+     * </p>
+     * <p>
+     *     You can specify a specific path for the index and dictionary file. If either set to {@code null} such
+     *     a file won't be generated.
+     * </p>
+     *
+     * @param fastaFile the output fasta file path.
+     * @param indexFile the path of the index file, if requested, {@code null} if none should be generated.
+     * @param dictFile the path of the dictFile, if requested, {@code null} if nono should be generated.
+     * @throws IllegalArgumentException if {@code fastaFile} is {@code null}.
+     * @throws IOException if such exception is thrown when accessing the output path resources.
+     */
+    public FastaReferenceWriter(final Path fastaFile, final Path indexFile, final Path dictFile)
+        throws IOException
+    {
+        this(fastaFile, DEFAULT_BASES_PER_LINE, indexFile, dictFile);
+    }
+
+    /**
+     * Creates a reference FASTA file writer.
+     * <p>
+     *     You can specify a specific path for the index and dictionary file. If either set to {@code null} such
+     *     a file won't be generated.
+     * </p>
+     *
+     * @param fastaFile the output fasta file path.
+     * @param indexFile the path of the index file, if requested, {@code null} if none should be generated.
+     * @param dictFile the path of the dictFile, if requested, {@code null} if nono should be generated.
+     * @throws IllegalArgumentException if {@code fastaFile} is {@code null} or {@code basesPerLine} is 0 or negative.
+     * @throws IOException if such exception is thrown when accessing the output path resources.
+     */
+    public FastaReferenceWriter(final Path fastaFile, final int basesPerLine, final Path indexFile, final Path dictFile)
+        throws IOException
+    {
+        // This code is a slight repeat of {@link #FastaReferenceWriter(OutputStream,int,OutputStream,OutputStream)
+        // for the sake of avoiding creating output if basesPerLine is invalid.
+        this.defaultBasePerLine = checkBasesPerLine(basesPerLine);
+        this.fastaStream = new CountingOutputStream(BucketUtils.createFile(Utils.nonNull(fastaFile).toString()));
```

Comment:
Would  be nice to being able to write as a gzipped/bgzipped file, because after my work in HTSJDK the framework would accept them

Reply:
Now that it is merged in HTSJDK, it would be also nice to generate the `GZIIndex` at some point...

Reply:
If the provided file name ends in *.gz, then it would be wrong to output a non-compressed file. In addition, if .fai and .dict is not provided, it would be nice to be able to output a compressed FASTA for an API user (I have some use cases).

Reply:
Full stack wise it would be wrong as is deceptive but the question is wether its this component's reposability to care of whether the name it is give makes sense for a non-compressed file. I would say that something else on top of it should be the one that cares. 

If a GATK's tool documentation says that .gz would result in a gizpped file then its code must make sure that this is the case but not the lower level component. 

In contrast some other tool in some other sofware package (when this sips thru to htsjdk) may also treat other extension as zipped ....perhaps .bgz, .Z ... is is rather difficult to provide some lower level component that would adapt to each given situation .... I think it makes more sense to delegate that responsability to a tool specific component that takes care to match the tools documentation.

As it is now, one can provide a gzipper outputStream that would take care of the compression and could possibly create the offset -> block index on the fly. 

The Path with explicit or implicit index Path constructors are provided for convenience as the would cover most common use cases. I'm a bit skeptical of the .dict file as for the most part is totally covered by the .fai which is the actual sequence look-up index but other components around simply can't work without a .dict file being present even they actually don't needed it.

Reply:
Most of the HTSJDK/Picard//GATK framework uses the extensions for correctly parse/write different formats and (b)gzip files. For example, if an user pass to this class a *.gz FASTA file and then try to read it with the framework, it would provide wrong results or fail because it would be assumed to be compressed when it is not. That's my concern.

I agree that automatically detecting compression by extension may be left by the API user, but that would require the usage of streams/channels in the library instead of `Path` or `File`. But that may be a lot of refactoring, I guess...